core/Pos/Crypto/HD.hs

```diff
@@ -1,4 +1,129 @@
+
+    It's not allowed to derive a public key from parent public key for
+    indicies which are greater than or equal to 2^31.
+    For these indicies secret key is required.
+    Such way to derive public keys is called "hardened derivation":
+    We also don't use it yet.
+    See more here
+    https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#private-parent-key--public-child-key
+
+    Note: all these functions don't require number of derivated level,
+    only credential information (public key or secret key with passphrase)
+    and index of a child.
+    Also we can derive public and secret keys for any arbitrary number of levels.
+
+* HD address payload
+    There were mentioned "account" and "address" entities,
+    but only addresses appear in the blockchain.
+    When we create a transaction, for each TxOut we specify address and coins.
+    So neither transaction nor blocks know anything about accounts and wallets.
+
+    We want to be able to track our addresses in the blockchain
+    to compute balance of wallet and accounts,
+    but as said in _General scheme_ section we aren't able
+    to determine a parent without any additional info.
+    So we can't determine account and wallet which address belongs to.
+
+    For each leaf let's store path from the root to this leaf along with an address.
+    So path is a list of indexes of derivation on each level.
+    To hide path from other parties we will encrypt it.
+    This encrypted derivation path is called @HDAddressPayload@
+    and stored in attributes of Address datatype.
+
+    Note: length of derivation path in our case is 2: @[account index, address index]@.
+
+* Encryption of payload
+    To encrypt derivation path let's use AEAD scheme using ChaCha20 and Poly1305.
+    To see more information go to https://tools.ietf.org/html/rfc7539.
+    Hash of root public key is used as a symmetric key for ChaChaPoly1305.
+    This hash called @HDPassphrase@ and function @deriveHDPassphrase@ generates
+    it by root public key.
+
+    Encryption of payload performs @packHDAddressAttr@ which takes @HDPassphrase@ and
+    derivation path and returns @HDAddressPayload@.
+    It serializes derivation path using @Bi@ instance for it and then encrypts produced bytes sequence.
+
+* Decryption of payload
+    To decrypt encrypted derivation path we have to derive @HDPassphrase@ again
+    from root public key and then try to decrypt @HDAddressPayload@.
+    If it's successfully decrypted then it implies address belongs to our tree
+    and, vice versa, address doesn't belong to our tree if decryption is failed.
+
+    Note: publishing of root public key gives other parties opportunity to reveal
+    all your addresses in the blockchain.
+    So it's not safe for the user to share his root public key (like he can do
+    it with a usual public key) because he loses his anonymity.
```

Comment:
"lose anonymity" doesn't seem to be a correct phrase. Use "it is unsafe to share root pk (like he can do) because it implies user deanonymization".